### PR TITLE
Adding class for generating all subsequences from a given List

### PR DESCRIPTION
**Pull request by alxkm**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

**Comment by vil02** [comment]
Is there a difference between `ArrayCombination` (currently under heavy refactoring in #5181) and your code? I like the fact that your code is generic.

File: src/main/java/com/thealgorithms/backtracking/Subsequence.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package com.thealgorithms.backtracking;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class generates all subsequences for a given list of elements using backtracking
+ */
+public final class Subsequence {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/backtracking/Subsequence.java:9]: What do you think about renaming this class. Perhaps `SubsequenceFinder` would be better.

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java
Comment (vil02) [inline src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:1]: Do you see some way to express your tests as _parametrized_? The challenge I see is to handle `List<List<String>>`, `List<List<Integer>>` etc. I cannot provide any productive suggestion, because I will be leaving my train in 10 minutes.

File: src/main/java/com/thealgorithms/backtracking/Subsequence.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     * We know that each state has exactly two branching
+     * It terminates when it reaches the end of the given sequence
+     *
+     * @param sequence all elements
+     * @param currentSubsequence current subsequence
+     * @param index current index
+     * @param allSubSequences contains all sequences
+     * @param <T> the type of elements which we generate
+     */
+    private static <T> void backtrack(List<T> sequence, List<T> currentSubsequence, int index, List<List<T>> allSubSequences) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/backtracking/Subsequence.java:42]:
```suggestion
    private static <T> void backtrack(List<T> sequence, List<T> currentSubsequence, final int index, List<List<T>> allSubSequences) {
```

File: src/main/java/com/thealgorithms/backtracking/Subsequence.java (hunk 1 above)
Comment (alxkm) [reply src/main/java/com/thealgorithms/backtracking/Subsequence.java]: I named it that way to be consistent with other classes like Permutation. 
However, you're right; it would be better to use SubsequenceFinder instead of Subsequence.

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java
Comment (alxkm) [reply src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:1]: I agree, all the tests are similar. I've refactored them into a parameterized format.

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.List;
+import java.util.stream.Stream;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class SubsequenceTest {
+
+    @ParameterizedTest
+    @MethodSource("getSubsequenceTestData")
+    void testSubsequences(SubsequenceTestData testData) {
```
Comment (vil02) [inline src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:15]: Please rename to `test<methodName>`.

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    static Stream<SubsequenceTestData> getSubsequenceTestData() {
+        return Stream.of(new SubsequenceTestData(new ArrayList<>(), List.of(List.of())), new SubsequenceTestData(List.of(1, 2), List.of(List.of(), List.of(2), List.of(1), List.of(1, 2))),
+            new SubsequenceTestData(List.of("A", "B", "C"), List.of(List.of(), List.of("C"), List.of("B"), List.of("B", "C"), List.of("A"), List.of("A", "C"), List.of("A", "B"), List.of("A", "B", "C"))),
+            new SubsequenceTestData(List.of(1, 2, 3), List.of(List.of(), List.of(3), List.of(2), List.of(2, 3), List.of(1), List.of(1, 3), List.of(1, 2), List.of(1, 2, 3))));
+    }
+
+    record SubsequenceTestData(List<Object> input, List<List<Object>> expected) {
+    }
```
Comment (vil02) [inline src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:27]: This is great!

I would suggest to rename it to:
```suggestion
    record TestCase(List<Object> input, List<List<Object>> expected) {
    }
```

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    @MethodSource("getSubsequenceTestData")
+    void testSubsequences(SubsequenceTestData testData) {
+        List<List<Object>> actual = SubsequenceFinder.generateAllSubsequences(testData.input());
+        assertIterableEquals(testData.expected(), actual);
+    }
+
+    static Stream<SubsequenceTestData> getSubsequenceTestData() {
+        return Stream.of(new SubsequenceTestData(new ArrayList<>(), List.of(List.of())), new SubsequenceTestData(List.of(1, 2), List.of(List.of(), List.of(2), List.of(1), List.of(1, 2))),
+            new SubsequenceTestData(List.of("A", "B", "C"), List.of(List.of(), List.of("C"), List.of("B"), List.of("B", "C"), List.of("A"), List.of("A", "C"), List.of("A", "B"), List.of("A", "B", "C"))),
+            new SubsequenceTestData(List.of(1, 2, 3), List.of(List.of(), List.of(3), List.of(2), List.of(2, 3), List.of(1), List.of(1, 3), List.of(1, 2), List.of(1, 2, 3))));
```
Comment (vil02) [inline src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:23]: What do you think about adding a test with duplicates?
```suggestion
            new SubsequenceTestData(List.of(1, 2, 3), List.of(List.of(), List.of(3), List.of(2), List.of(2, 3), List.of(1), List.of(1, 3), List.of(1, 2), List.of(1, 2, 3))),
            new SubsequenceTestData(List.of(2, 2), List.of(List.of(), List.of(2), List.of(2), List.of(2, 2))));
```

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.stream.Stream;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class SubsequenceTest {
+
+    @ParameterizedTest
+    @MethodSource("getSubsequenceTestData")
+    void testSubsequences(SubsequenceTestData testData) {
+        List<List<Object>> actual = SubsequenceFinder.generateAllSubsequences(testData.input());
```
Comment (vil02) [inline src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:16]:
```suggestion
        final var actual = SubsequenceFinder.generateAllSubsequences(testData.input());
```

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        assertIterableEquals(testData.expected(), actual);
+    }
+
+    static Stream<SubsequenceTestData> getSubsequenceTestData() {
```
Comment (vil02) [inline src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java:20]:
```suggestion
    static Stream<SubsequenceTestData> getTestCases() {
```

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertIterableEquals;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class SubsequenceTest {
```
[the remark this answers is not in the document]
Comment (alxkm) [reply src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java]: Thanks. The test name should also be updated as a class.

File: src/main/java/com/thealgorithms/backtracking/SubsequenceFinder.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    /**
+     * Find all subsequences of given list using backtracking
+     *
+     * @param sequence a list of items on the basis of which we need to generate all subsequences
+     * @param <T> the type of elements in the array
+     * @return a list of all subsequences
+     */
+    public static <T> List<List<T>> generateAllSubsequences(List<T> sequence) {
```
[the remark this answers is not in the document]
Comment (alxkm) [reply src/main/java/com/thealgorithms/backtracking/SubsequenceFinder.java]: Definitely, since the existing name is in some sense a duplicate of the class name.

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java (hunk 4 above)
Comment (alxkm) [reply src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java]: Thanks.

File: src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java (hunk 5 above)
Comment (alxkm) [reply src/test/java/com/thealgorithms/backtracking/SubsequenceTest.java]: Yes, ensuring the duplicate test is included is important. Thanks.

File: src/main/java/com/thealgorithms/backtracking/SubsequenceFinder.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+     * It terminates when it reaches the end of the given sequence
+     *
+     * @param sequence all elements
+     * @param currentSubsequence current subsequence
+     * @param index current index
+     * @param allSubSequences contains all sequences
+     * @param <T> the type of elements which we generate
+     */
+    private static <T> void backtrack(List<T> sequence, List<T> currentSubsequence, final int index, List<List<T>> allSubSequences) {
+        if (index == sequence.size()) {
```
Comment (vil02) [inline src/main/java/com/thealgorithms/backtracking/SubsequenceFinder.java:43]:
```suggestion
        assert index <= sequence.size();
        if (index == sequence.size()) {
```

**Review by vil02 — APPROVED** [review]
Looks good. Dzięki!
